File: backends/node-nestjs/src/seed.ts
import { readdir, readFile } from "node:fs/promises";
import path from "node:path";
import { config } from "./config.js";
import { pool } from "./db.js";
import { logEvent } from "./logging.js";

/**
 * SPEC rule 12: import the JSON seed files from `spec/messages` (language =
 * filename), inserting only `(key, language)` pairs that don't exist yet —
 * runtime edits by admins survive restarts. Seed inserts are not moderator
 * actions, so they are deliberately not audited.
 */
export async function seedMessages(): Promise<void> {
  const dir = config.seedMessagesDir;
  const files = (await readdir(dir).catch(() => null))?.filter((file) => file.endsWith(".json")).sort();
  if (!files) {
    throw new Error(
      `Message seed directory not found: ${dir} — set SEED_MESSAGES_DIR to the spec/messages directory`,
    );
  }
  for (const file of files) {
    const language = path.basename(file, ".json");
    const entries = JSON.parse(await readFile(path.join(dir, file), "utf8")) as Record<string, string>;
    const keys = Object.keys(entries);
    const result = await pool.query(
      `insert into messages (id, key, language, text, created_at, updated_at)
       select gen_random_uuid(), key, $1, text, now(), now()
       from unnest($2::text[], $3::text[]) as seed(key, text)
       on conflict (key, language) do nothing`,
      [language, keys, keys.map((key) => entries[key])],
    );
    const inserted = result.rowCount ?? 0;
    logEvent(
      "info",
      "message_seed_imported",
      "success",
      `Message seed '${language}': ${inserted} inserted, ${keys.length - inserted} already present`,
      { language, inserted, skipped: keys.length - inserted },
    );
  }
}

File: backends/go-echo/internal/messages/seed.go
package messages

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/google/uuid"

	"github.com/kamkie/stackverse/backends/go-echo/internal/logx"
	"github.com/kamkie/stackverse/backends/go-echo/internal/store"
)

// Seed implements SPEC rule 12: import the JSON files from spec/messages
// (language = filename), inserting only (key, language) pairs that don't exist
// yet — runtime edits by admins survive restarts. Seed inserts are not
// moderator actions, so they are deliberately not audited.
func (s *Store) Seed(ctx context.Context, dir string, logger *slog.Logger) error {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return fmt.Errorf("message seed directory not found: %s — set SEED_MESSAGES_DIR to the spec/messages directory: %w", dir, err)
	}
	names := make([]string, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".json") {
			names = append(names, entry.Name())
		}
	}
	sort.Strings(names)
	for _, name := range names {
		if err := s.seedLanguage(ctx, dir, name, logger); err != nil {
			return err
		}
	}
	return nil
}

func (s *Store) seedLanguage(ctx context.Context, dir, name string, logger *slog.Logger) error {
	language := strings.TrimSuffix(name, ".json")
	raw, err := os.ReadFile(filepath.Join(dir, name))
	if err != nil {
		return err
	}
	var texts map[string]string
	if err := json.Unmarshal(raw, &texts); err != nil {
		return fmt.Errorf("message seed %s: %w", name, err)
	}
	keys := make([]string, 0, len(texts))
	for key := range texts {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	transaction, err := s.pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = transaction.Rollback(ctx) }()

	now := store.NowUTC()
	inserted := 0
	for _, key := range keys {
		tag, err := transaction.Exec(ctx,
			`insert into messages (id, key, language, text, description, created_at, updated_at)
			 values ($1, $2, $3, $4, null, $5, $5)
			 on conflict (key, language) do nothing`,
			uuid.New(), key, language, texts[key], now)
		if err != nil {
			return err
		}
		inserted += int(tag.RowsAffected())
	}
	if err := transaction.Commit(ctx); err != nil {
		return err
	}
	logx.Event(ctx, logger, slog.LevelInfo, "message_seed_imported", "success",
		fmt.Sprintf("Message seed '%s': %d inserted, %d already present", language, inserted, len(keys)-inserted),
		slog.String("language", language),
		slog.Int("inserted", inserted),
		slog.Int("skipped", len(keys)-inserted),
	)
	return nil
}
